src/relentless/optimize/objective.py

```diff
@@ -374,6 +383,286 @@ def compute(self, variables, directory=None):
+
+                bonded_exclusions = self.potentials.pair.exclusions
+                if bonded_exclusions is not None:
+                    if (
+                        "1-2" in bonded_exclusions
+                        and snap.bonds.N != 0
+                        and len(neighbors[:]) > 0
+                    ):
+                        bonds = numpy.vstack(
+                            [snap.bonds.group, numpy.flip(snap.bonds.group, axis=1)],
+                        )
+
+                        bond_exclusion_filter = EnsembleAverage._cantor_pairing(
+                            self, bonds, neighbors
+                        )
+
+                        neighbors.filter(bond_exclusion_filter)
+                    if (
+                        "1-3" in bonded_exclusions
+                        and snap.angles.N != 0
+                        and len(neighbors[:]) > 0
+                    ):
+                        angles = numpy.vstack(
+                            [snap.angles.group, numpy.flip(snap.angles.group, axis=1)],
+                        )
+
+                        angle_exclusion_filter = EnsembleAverage._cantor_pairing(
+                            self, angles[:, (0, -1)], neighbors
+                        )
+
+                        neighbors.filter(angle_exclusion_filter)
+                    if (
+                        "1-4" in bonded_exclusions
+                        and snap.dihedrals.N != 0
+                        and len(neighbors[:]) > 0
+                    ):
+                        dihedrals = numpy.vstack(
+                            [
+                                snap.dihedrals.group,
+                                numpy.flip(snap.dihedrals.group, axis=1),
+                            ],
+                        )
+
+                        dihedral_exclusion_filter = EnsembleAverage._cantor_pairing(
+                            self, dihedrals[:, (0, -1)], neighbors
+                        )
+
+                        neighbors.filter(dihedral_exclusion_filter)
+
+                # pair contributions to the gradient
+                for i in snap.particles.types:
+                    for j in snap.particles.types:
+                        filter_j_gt_i = neighbors[:, 1] > neighbors[:, 0]
+                        neighbors.filter(filter_j_gt_i)
+
+                        filter_ij = numpy.logical_and(
+                            type_masks[i][neighbors[:, 0]],
+                            type_masks[j][neighbors[:, 1]],
+                        )
+
+                    neighbors.filter(filter_ij)
+
+                    distances = neighbors.distances[filter_ij]
+
+                    for var in variables:
+                        rs = self.potentials.pair.linear_space
+                        dus = self.potentials.pair.derivative((i, j), var)
+                        dudvar = math.AkimaSpline(rs, dus)
+                        gradient[var] += numpy.sum(dudvar(distances)) / len(traj)
```

Comment:
Can't this be done by specifying x directly, rather than passing through an intermediate interpolation?
```suggestion
                        gradient[var] += numpy.sum(self.potentials.pair.derivative((i, j), var, x=distances))
```
Side note: I would normalize the whole thing by the number of frames at the end

Same comments apply below.